New pack: Hackuity

Rogdham opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
<!--
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**
-->

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
None

## Description
This new pack provides an integration with Hackuity, a vulnerability management solution.

## Screenshots

<details>
<summary>Configuration page</summary>

![Configuration](https://user-images.githubusercontent.com/3994389/149543057-d4e28cf5-d981-4d47-869a-47c8afcf67b6.png)
![About page](https://user-images.githubusercontent.com/3994389/149543101-098a4c31-0cb5-432a-87ed-99f91627eedf.png)

</details>


<details>
<summary>hackuity-search-findings command</summary>

![!hackuity-search-findings](https://user-images.githubusercontent.com/3994389/149543579-594e5250-1d76-4605-8bd9-cbeae038f58f.png)

</details>

<details>
<summary>hackuity-search-vulndb-vulnerabilities command</summary>

![!hackuity-search-vulndb-vulnerabilities](https://user-images.githubusercontent.com/3994389/149544605-2caabcef-c1ae-4df3-a99b-a9445ee11c82.png)

</details>

## Minimum version of Cortex XSOAR
- [x] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=17106) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=17106) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=17106) it.</sub>

Rogdham commented:
MyPy lint fails due to https://github.com/demisto/demisto-sdk/issues/1831

merit-maita commented:
@Rogdham Thanks for the contribution, please fill the [registration form](https://forms.gle/XDfxU4E61ZwEESSMA) so we can  go ahead with the review.

Rogdham commented:
@merit-maita thank you for your reply, it should be done now!

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+    verify_certificate = not demisto.params().get("insecure", False)
+    proxy = demisto.params().get("proxy", False)
+
+    commands: Dict[str, Callable[[Client, Dict[str, str]], CommandResults]] = {
```
```suggestion
    commands = {
```

Rogdham replied:
I have added that type otherwise MyPy thinks it is the following:
```
builtins.dict[builtins.str*, builtins.object*]
```

…and fails on a later line with this error:

```
/Packs/Hackuity/Integrations/Hackuity/Hackuity.py:673:28: error:
"object" not callable  [operator]
                return_results(commands[command](client, demisto.args()))
```

Please feel free to suggest a better solution for MyPy!

merit-maita replied:
its okay you can keep it then

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+    client: Client, args: Dict[str, Any]
+) -> CommandResults:
+    widget_id = args["widget_id"]
+    outputs = client.fetch_dashboard_widget_data(widget_id=widget_id)
```
in case there's no widget with the given id, it would return an 404 error, in that case, we'd want to return a readable output stating that and not fail the command.

Rogdham replied:
Good catch!

Changed to the following, tell me if it's not the best way to display the error
![screenshot-2022-01-26-18:29:05](https://user-images.githubusercontent.com/3994389/151215219-a9042117-98ed-4346-bd14-6c77cc6ff9e5.png)

merit-maita replied:
looks perfect

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+def hackuity_dashboard_widgets_command(
+    client: Client, args: Dict[str, Any]
+) -> CommandResults:
+    widgets = client.fetch_dashboard_widgets()
```
in case non is returned (no dashboards) return a readable output stating that, do not fail the command

Rogdham replied:
By design the dashboard with the user id (the one we are requesting in `fetch_dashboard_widgets`) will always exist; in case there are no widgets `fetch_dashboard_widgets` will return an empty list, I think it is ok

merit-maita replied:
okay, sounds good

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+    vuln_type = args.get("vuln_type")
+    trs_min = none_or_apply(args.get("trs_min"), int)
+    trs_max = none_or_apply(args.get("trs_max"), int)
+    findings = client.fetch_findings(
```
make sure in case there's non, a readable output is returned an not an error

Rogdham replied:
If there are no findings, `fetch_findings` will return an empty list, I think this is ok

![screenshot-2022-01-26-18:17:26](https://user-images.githubusercontent.com/3994389/151213336-b4c8199d-b714-4139-98e3-b6c3578c6c46.png)

merit-maita replied:
sounds good too!

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+    vuln_type = args.get("vuln_type")
+    trs_min = none_or_apply(args.get("trs_min"), int)
+    trs_max = none_or_apply(args.get("trs_max"), int)
+    aggfindings = client.fetch_aggfindings(
```
make sure in case there's non, a readout output is returned and not an error

Rogdham replied:
If there are no aggfindings, `fetch_aggfindings` will return an empty list, I think this is ok

![screenshot-2022-01-26-18:19:02](https://user-images.githubusercontent.com/3994389/151213549-5da56bab-f9d1-4b8b-a727-d0136cd2ad40.png)

merit-maita replied:
perfect

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.yml:
```diff
@@ -0,0 +1,228 @@
+      description: The ID of the vulnerability
+      type: string
+    - contextPath: Hackuity.Vulnerabilities.Seen.First
+      description: The date of the first time the vulnerability has been seen
```
since you have a specific format, add an example for this date

Rogdham replied:
The output date is specifically formatted with `DATE_FORMAT` which I thought was the way dates are represented inside Cortex XSOAR? e.g. `2021-03-03T07:56:07Z`

Tell me if that's not the case, so that I can change the date formatting so that is it the one used by default inside Cortex XSOAR.

merit-maita replied:
yeah its okay, just added to the description as an example, that's it

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/README.md:
```diff
@@ -0,0 +1,175 @@
+
+#### Context Output
+
+There is no context output for this command.
```
why dont you have a command example with returned data to any of the commands? it should be provided.

Rogdham replied:
Done

merit-maita replied:
where? the file is not updated

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity_test.py:
```diff
@@ -0,0 +1,48 @@
+            "search_vulnerabilities",
+            id="search_vulnerabilities",
+        ),
+    ],
```
what about the other commands?

Rogdham replied:
The `hackuity-dashboard-data` and `hackuity-dashboard-widgets` commands are provided in this first version of the integration as best-effort.

The `hackuity-search-provider-vulnerabilities` and `hackuity-search-vulndb-vulnerabilities` commands are very similar so I thought it was enough to add a test for only one of them, but tell me if you want me to add a test for the other one.

merit-maita replied:
there should be a test for each command even it some are similar

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/command_examples.txt:
```diff
@@ -0,0 +1,3 @@
+!hackuity-search-findings asset_name=example.com
+!hackuity-search-provider-vulnerabilities asset_name=example.com
+!hackuity-search-vulndb-vulnerabilities asset_name=example.com
```
command examples used in building readme, should be example that would return reasonable and correct results, which i assume that example.com does not really work in that case

Rogdham replied:
Done

merit-maita replied:
where do you mean its done? the file is not updated

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.yml:
```diff
@@ -0,0 +1,228 @@
+  type: python
+  subtype: python3
+  dockerimage: demisto/python3:3.10.1.25933
+fromversion: 5.5.0
```
since its a new integration its preferred to be supported from 6.0.0, is there a specific reason its set to be 5.5.0?

Rogdham replied:
Done

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+
+""" CONSTANTS """
+
+__version__ = "1.0.0"
+
+DATE_FORMAT = r"%Y-%m-%dT%H:%M:%SZ"
+
+DEMISTO_VERSION = get_demisto_version_as_str()
+CONTENT_PACK = f"Hackuity/{__version__}"
+INTEGRATION = get_integration_name()
+USER_AGENT = f"XSOAR/{DEMISTO_VERSION} - {CONTENT_PACK} - {INTEGRATION}"
```
```suggestion
DATE_FORMAT = r"%Y-%m-%dT%H:%M:%SZ"

USER_AGENT = f"XSOAR/{get_demisto_version_as_str()} - Hackuity/1.0.0 - {get_integration_name()}"
```

Rogdham replied:
Done

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+            return_results(commands[command](client, demisto.args()))
+        else:
+            raise NotImplementedError(
+                f"{command} is not an existing {INTEGRATION} command"
```
```suggestion
                f"{command} is not an existing command"
```

Rogdham replied:
Done

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+                f"/widgets/{widget_id}/data"
+            ),
+        )
+        return json_get_value(response, ["payload", None])
```
if your trying to get the json object from the response why dont you use json() method?

Rogdham replied:
Because my `http_request` passes a `resp_type` of value `json` by default to `BaseClient._http_request`, the `response` object will already be json-decoded.

---

The `None` value in the `path` arg of `json_get_value` means “get the first value you find”.

The response returned by the server looks like this:
```json
{
  "kind": "hy#widgetData",
  "payload": {
    "assetsOverview": "redacted for clarity"
  },
  "type": "ASSETS_OVERVIEW"
}
```

And we are only looking to return the `redacted for clarity` part.

merit-maita replied:
cool, we'll talk about it in the demo

merit-maita commented on Packs/Hackuity/Integrations/Hackuity/Hackuity.py:
```diff
@@ -0,0 +1,694 @@
+            ),
+        )
+        return [
+            json_extract(
+                widget,
+                {
+                    "id": ["id"],
+                    "params": ["params", None],
+                    "type": ["type"],
+                },
+            )
+            for widget in response.get("widgets", [])
```
if you're trying to get the json object of each widget, why dont you just return it as it is from the api, or use an already existing methods.

Rogdham replied:
Because my `http_request` passes a `resp_type` of value `json` by default to `BaseClient._http_request`, the `response` object will already be json-decoded.


---

The `None` value in the `path` arg of `json_get_value` means “get the first value you find”.

The response returned by the server looks like this (some extra fields have been removed):
```json
{
  "kind": "hy#dashboard",
  "widgets": [
    {
      "id": "abc123",
      "params": {
        "overallScore": "redacted for clarity #1"
      },
      "type": "OVERALL_SCORE"
    },
    {
      "id": "def456",
      "params": {
        "scoreMeter": "redacted for clarity #2"
      },
      "type": "SCORE_METER"
    }
  ]
}
```

And we are only looking to return the following (note that we removed the nesting for the `params` keys):
```json
[
  {
    "id": "abc123",
    "params": "redacted for clarity #1",
    "type": "OVERALL_SCORE"
  },
  {
    "id": "def456",
    "params": "redacted for clarity #2",
    "type": "SCORE_METER"
  }
]
```

merit-maita reviewed (changes requested):
Thanks for the hard work
please make sure to pass the checks of the pr.

Rogdham commented:
Thank you for the review @merit-maita! All your points should now have been fixed or at least answers in a comment.

I think all checks that are supposed to pass on this PR should be good now.

Except that [CircleCI is currently having an outage](https://status.circleci.com/) so the *Run Unit Testing And Lint* is failing because if it.

---

Sorry for the several pushes, this was me trying to fix the [MyPy failing on Python 3.10](https://github.com/demisto/demisto-sdk/issues/1831) issue:
 - I initially rebased on `origin/master`, but I see now that is was not a good idea because the base branche of this PR has been changed to `contrib/Rogdham_hackuity` (instead of `origin/master`)
 - I canceled the rebase and added a commit to change version of `demisto-sdk` used is 1.5.8 and not 1.5.7 :arrow_forward: this change is already on `origin/master`

:arrow_forward: now the commits in this branch are clearer: your last review was on 5c29226c6b0d32ab5edc0c9528d100cdc2a97a7b and I pushed two commits on top: one to address your review, and one to fix the MyPy issue.

Sorry for the noise :disappointed:

merit-maita commented:
@Rogdham first of all, thanks for the hard work and for addressing all of the comments 
so i need to schedule a demo, please let me know when it works for you,
and for the lint failing check, please pull again and push, i dont have permissions to fix it.

Rogdham commented:
> for the lint failing check, please pull again and push, i dont have permissions to fix it.

Done

> please let me know when it works for you

@merit-maita can I let you send me an email (my address is in the commits / CLA signed / Contributor registration form) with some times that works for you this week ([my local time](https://time.is/Paris))? :pray:

Rogdham commented:
@merit-maita following the demo, here are the changes I made:
1. Remove column `attributes` from markdown table in command `hackuity-search-provider-vulnerabilities`
2. Add all commands to `command_examples.txt`, generate `README.md`, take care of `Hackuity.yml`
3. Filter the responses inside each command instead of in the client
    - Because of that, the `json_xxx` helpers have been refactored and merged into a single `remap_item` helper
    - Also I noticed that the library `funcy` is available so I used it to avoid reinventing the `get_in`/`set_in` helpers
4. Add tests for all missing commands

Unfortunately this took a lot of time but I think everything you mentioned both inside PR reviews and during the demo have been taken care of! :pray:

merit-maita commented:
> @merit-maita following the demo, here are the changes I made:
> 
> 1. Remove column `attributes` from markdown table in command `hackuity-search-provider-vulnerabilities`
> 2. Add all commands to `command_examples.txt`, generate `README.md`, take care of `Hackuity.yml`
> 3. Filter the responses inside each command instead of in the client
>    
>    * Because of that, the `json_xxx` helpers have been refactored and merged into a single `remap_item` helper
>    * Also I noticed that the library `funcy` is available so I used it to avoid reinventing the `get_in`/`set_in` helpers
> 4. Add tests for all missing commands
> 
> Unfortunately this took a lot of time but I think everything you mentioned both inside PR reviews and during the demo have been taken care of! 🙏

@Rogdham Thank you for the great effort you made!!, please take a look at the checks failing so it could be merged.
thanks again.

Rogdham commented:
Hello @merit-maita , the test were failing because they were some changes on `origin/master`.

I changed the target of the PR to `origin/master` and rebased my PR, but as you can see now other tests are failling because the PR is not based on `contrib/Rogdham_Hackuity` anymore.

So I'm confused, could you advise how to proceed?

Rogdham commented:
As discussed:
 - Put my contribution in a single commit on top of `contrib/Rogdham_Hackuity`
 - Then merge `master` on top of that

Feel free to write back to me if you want things to be done differently.
